src/riak_kv_query_buffer.erl

```diff
@@ -0,0 +1,599 @@
+        {terms, term_accumulator()} |
+        {raw_terms, rawterm_accumulator()} |
+        {term_with_rawcount, countby_aggregator()} |
+        {term_with_count, countby_aggregator()}.
+-type reply_fun()
+    :: fun((reply_type()|ping) -> ok).
+
+-type buffer_agg()
+    ::
+        #key_agg{} |
+        #termcount_agg{} |
+        #termkeycount_agg{}.
+
+-type buffer() :: #buffer{}.
+
+-export_type([reply_type/0, reply_fun/0, buffer/0]).
+
+-spec new(
+    {pos_integer(), non_neg_integer()},
+    riak_kv_query:accumulation_option(),
+    reply_fun())
+        -> buffer().
+new(Size, T, ReplyFun) when T == keys ->
+    new_buffer(Size, ReplyFun, keys, none, T);
+new(Size, T, ReplyFun) when T == raw_keys ->
+    new_buffer(Size, ReplyFun, raw_keys, none, T);
+new(Size, T, ReplyFun) when T == count->
+    new_buffer(Size, ReplyFun, raw_keys, #key_agg{}, T);
+new(Size, T, ReplyFun) when T == raw_count ->
+    new_buffer(Size, ReplyFun, none, none, T);
+new(Size, T, ReplyFun) when T == terms->
+    new_buffer(Size, ReplyFun, terms, none, T);
+new(Size, T, ReplyFun) when T == raw_terms->
+    new_buffer(Size, ReplyFun, raw_terms, none, T);
+new(Size, T, ReplyFun) when T == term_with_rawcount->
+    new_buffer(Size, ReplyFun, terms, #termcount_agg{}, T);
+new(Size, T, ReplyFun) when T == term_with_count ->
+    new_buffer(Size, ReplyFun, terms, #termkeycount_agg{}, T).
+
+new_buffer({BufferSize, JitterSize}, ReplyFun, AccType, InitAgg, Type)
+        when BufferSize >= ?MIN_BUFFER, JitterSize =< BufferSize ->
+    {RawKeyAcc, KeyAcc, RawTermAcc, TermAcc} =
+        case AccType of
+            none -> {none, none, none, none};
+            raw_keys -> {[], none, none, none};
+            keys -> {none, [], none, none};
+            terms -> {none, none, none, []};
+            raw_terms -> {none, none, [], none}
+        end,
+    ActualBufferSize =
+        case JitterSize of
+            0 ->
+                BufferSize;
+            JS when JS > 0 ->
+                (BufferSize - (JS div 2)) + rand:uniform(JS)
+        end,
+    #buffer{
+        max_size = ActualBufferSize,
+        reply_fun = ReplyFun,
+        rawkey_acc = RawKeyAcc,
+        key_acc = KeyAcc,
+        rawterm_acc = RawTermAcc,
+        term_acc = TermAcc,
+        agg = InitAgg,
+        type = Type
+    }.
+
+min_buffer() -> ?MIN_BUFFER.
+
+-spec add(binary()|{binary(), binary()}, buffer()) -> buffer().
+add(Key, #buffer{max_size = MS, count = MS} = Buffer) ->
+    add(Key, merge(Buffer));
+add(_Key,
+    #buffer{
+        rawkey_acc = none, key_acc = none, term_acc = none, rawterm_acc = none,
+        count = C
+    } = B) ->
+    B#buffer{count = C + 1};
+add(Key, #buffer{rawkey_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Key) ->
+    Buffer#buffer{rawkey_acc = [Key|A], count = C + 1};
+add(Key, #buffer{key_acc = A, count = C} = Buffer)
+        when A =/= none, is_binary(Key) ->
+    Buffer#buffer{key_acc = [{Key}|A], count = C + 1};
```

Comment:
Why these `{ .. }` in the key_acc and later in term_acc. Would it not make sense to do the same as for raw key and raw term and then use the fact that you know what the type is later on?
I'm probably missing something here.

Reply:
This is related to the use of ets for sorting.  Treating them this way means that they cna be handled the same when adding them into ets.

https://github.com/OpenRiak/riak_kv/commit/85f32c5d2c42f769ddfbdb705087683737d0f0cd

For raw key and raw term, they're not going to be inserted into ets for sorting, and so this doesn't matter.

I can't remember the details, but I do recall it was very deliberate.

Reply:
Looking back at other commit references, the issue is that the tuple wrapping is required for ets, but also riak_kv_query_server becomes the bottleneck when its sorting.  This is why the tuple wrapping happens at the buffer - so that an additional lists:map is not required to be done on the query_server